### PR TITLE
Preserve float64 sparse xent tail gradients

### LINKED ISSUE #126636: tf.nn.sparse_softmax_cross_entropy_with_logits loses a finite first-derivative signal for float64 input
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.sparse_softmax_cross_entropy_with_logits` returns zero for one logit-gradient component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

Sparse softmax cross entropy is invariant to adding the same constant to all logits, so the gradient with respect to logits should sum to zero. In the reproducer below, the expected gradient is approximately `[-5.551115123125776e-17, 5.551115123125776e-17]`.

However, TensorFlow returns `[[0.0, 5.551115123125776e-17]]`. The first gradient component is incorrectly collapsed to zero, and the resulting logit gradient no longer preserves the expected zero-sum relation.

Expected behavior?

The first-order autodiff result for `tf.nn.sparse_softmax_cross_entropy_with_logits` should preserve the finite, normal, nonzero `float64` gradient component. For the given logits and sparse label, the gradient should be close to `[[-5.551115123125776e-17, 5.551115123125776e-17]]`, and the gradient sum should be zero.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

dtype = tf.float64

expected = [-5.551115123125776e-17, 5.551115123125776e-17]

logits = tf.Variable([[37.42994775023705, 0.0]], dtype=dtype)
labels = tf.constant([0], dtype=tf.int32)

with tf.GradientTape() as tape:
    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
        labels=labels,
        logits=logits,
    )
    objective = tf.reduce_sum(loss)

actual = tape.gradient(objective, logits)
grad_sum = tf.reduce_sum(actual)

print("loss:")
print(loss.numpy().tolist())
print("actual:")
print(actual.numpy().tolist())
print("expected:")
print(expected)
print("gradient sum:")
print(float(grad_sum))

if actual.numpy()[0, 0] == 0.0 and expected[0] != 0.0:
    print("BUG REPRODUCED: tf.nn.sparse_softmax_cross_entropy_with_logits loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
loss:
[0.0]
actual:
[[0.0, 5.551115123125776e-17]]
expected:
[-5.551115123125776e-17, 5.551115123125776e-17]
gradient sum:
5.551115123125776e-17
BUG REPRODUCED: tf.nn.sparse_softmax_cross_entropy_with_logits loses a finite first-derivative signal
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
## Summary

- derive the labeled float64 gradient from the sum of the other class probabilities
- preserve finite tail gradients when the labeled probability rounds to one
- leave lower-precision and GPU execution paths unchanged

Fixes #126636.

## Testing

- `bazel build //tensorflow/core/kernels:sparse_xent_op`
- added graph/eager regression coverage for `int32` and `int64` labels

### CODE REVIEW

**Comment by GodlyDonuts** [comment]
@gemini-code-assist /review

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for the PR to preserve float64 sparse xent tail gradients!

The GitHub CI tests failed on both compilation jobs (CPU and CUDA) as well as the PyLint check. 

1. **Compilation Failures**: It looks like `is_same_v` and `if constexpr` were introduced in `sparse_xent_op.h` which causes compilation errors on some of the upstream presubmit CI toolchains. Could you please fallback to `std::is_same<T, double>::value` and avoid `if constexpr` to ensure it successfully builds across all CI compilers? We've verified this caused the failure seen in Linux CPU and CUDA builds.

2. **PyLint Failure**: PyLint failed on `tensorflow/python/kernel_tests/sparse_ops/sparse_xent_op_test_base.py`. Could you fix the lint errors on your PR to satisfy the python linters in CI?

Thanks again, looking forward to the updates!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi, thank you for the updates! The unit tests look good and now pass the test suite, and the CI is green.
However, our internal strict dependency builds caught a missing dependency in the `BUILD` file. Could you please add `"//tensorflow/python/ops:nn_ops_gen"` to the `deps` of `sparse_xent_op_test_base` in `tensorflow/python/kernel_tests/sparse_ops/BUILD`?

```diff
--- a/tensorflow/python/kernel_tests/sparse_ops/BUILD
+++ b/tensorflow/python/kernel_tests/sparse_ops/BUILD
@@ -391,6 +391,7 @@ py_library(
         "//tensorflow/python/ops:gradient_checker_v2",
         "//tensorflow/python/ops:nn_grad",
         "//tensorflow/python/ops:nn_ops",
+        "//tensorflow/python/ops:nn_ops_gen",
         "//tensorflow/python/platform:client_testlib",
     ],
 )
```

Once this is added, the PR will be ready to merge!

**Review by dmiltr3 — APPROVED** [review]
Thanks for making the requested adjustments! I've verified the fix by running our internal automated test suite and running the GitHub CI. The upstream compilation errors have been resolved, and internal tests pass cleanly.

I will go ahead and approve this PR. Thanks again for your contribution!
